Embedded JS: Conditionally inject jQuery

stsewd opened the pull request:
For injecting the flyout menu we are relying on a script from our theme, that script depends on jQuery, so it needs to be present for the flyout menu to work.

https://github.com/readthedocs/readthedocs.org/blob/67c44a1f1b27dcc66bd14520d019459fcf86354d/readthedocs/core/static-src/core/js/doc-embed/sphinx.js#L43-L59

We have migrated most of our code to vanilla JS,
but this piece isn't simple to migrate.

ericholscher reviewed (approved):
This is a good addition, and something I've been wanting for a while now 👍

benjaoming commented on readthedocs/core/static-src/core/js/readthedocs-doc-embed.js:
```diff
@@ -6,11 +6,36 @@ const sphinx = require('./doc-embed/sphinx');
+    console.log("JQuery not found. Injecting.");
+    var script = document.createElement("script");
+    script.type = "text/javascript";
+    script.src = "https://code.jquery.com/jquery-3.6.3.min.js";
```
Should this also be loaded from a readthedocs.io domain because of CSP?

agjohnson replied:
Yeah agreed with @benjaoming, for CSP and CORS related headaches, this should be from our own domain.

I'd probably say we should just inject our existing jQuery vendored bundle, which is compiled using Gulp/Bower. Our current vendored jQuery is 2.0 however, but the API hasn't changed too much as well: https://github.com/readthedocs/readthedocs.org/blob/main/bower.json#L17

stsewd replied:
this is loaded from docs domains, we don't have a CSP there, but +1 on serving this from `/_/static/`. About the version, guess, anything that works with https://github.com/readthedocs/sphinx_rtd_theme/blob/0.3.1/js/theme.js should be fine. But since our theme was depending on the jquery version from sphinx not sure if there would be any problems, but running locally should be enough to test it I guess

benjaoming replied:
If we look, we can probably find security-related issues addressed in jQuery updates, so it's probably best to stick to the latest version. That is the approach of `sphinxcontrib-jquery` as well, so thus the two approaches can align with the same jQuery version.

For `sphinxcontrib-jquery`, I think the approach has also been that if someone steps up and wants configuration of the jQuery version, it will be implemented... but on-demand. See also: https://github.com/sphinx-contrib/jquery/blob/master/sphinxcontrib/jquery/__init__.py

benjaoming replied:
The jQuery version from Sphinx was just which-ever latest version that happened to be vendored when that Sphinx version was released: https://github.com/sphinx-doc/sphinx/commits/v5.3.0/sphinx/themes/basic/static/jquery.js

agjohnson replied:
I'm guessing we're also probably okay upgrading the jQuery version installed via Bower, we don't do anything particularly interesting in the existing templates/JS.

stsewd replied:
```
Unable to find a suitable version for jquery, please choose one by typing one of the numbers below:
    1) jquery#~1.8.1 which resolved to 1.8.3 and is required by jquery-ui#1.8.23
    2) jquery#3.6.3 which resolved to 3.6.3 and is required by readthedocs
    3) jquery#>=1.5 which resolved to 3.6.3 and is required by jquery.payment#1.3.3
```

Don't think we are using the other plugins. I went ahead and wrote `2!`, but I don't see the new files being copied to media, hmm, do we have a script for copying the files? :thinking:

stsewd replied:
Traveled back in time to find `npm run vendor`

https://github.com/readthedocs/readthedocs.org/blob/3.11.0/docs/development/front-end.rst

agjohnson replied:
I don't recall what we even use jquery-ui for, but the latest version of jquery-ui is a minor version bump and supports jQuery =< 3.6.

Good question on the static files. Vendored libraries are processed with Gulp here:
https://github.com/readthedocs/readthedocs.org/blob/main/gulpfile.js#L193

And are output here:
https://github.com/readthedocs/readthedocs.org/tree/main/readthedocs/static/vendor

agjohnson replied:
Ah missed your reply until reload. I thought `gulp build` built all assets, but the command is separate, `gulp vendor`.

stsewd commented on readthedocs/core/static-src/core/js/readthedocs-doc-embed.js:
```diff
@@ -5,12 +5,43 @@ const footer = require('./doc-embed/footer.js');
+    script.type = "text/javascript";
+    script.src = rtd.proxied_static_path + "vendor/jquery.js";
+    script.onload = function () {
+        // Set jQuery to its expected globals.
+        window.$ = require("jquery");
+        window.jQuery = window.$;
```
Not really sure if this is the way of doing it... but just adding the script from vendor/jquery.js doesn't add the global variables :/ /cc @agjohnson

agjohnson replied:
Yup, this is normal. Relying on every module to be a global is an antipattern that doesn't work well with various forms of module importing -- commonJS, AMD, ES6 module imports, etc. Modern JS would tend to use jQuery as a module instead.

Your change is about what I'd do.

stsewd commented:
Ok, I have updated this PR to:

- Inject jquery from the same domain as the docs using our proxied static files endpoint.
- Update the jquery version to the latest released one.
- Updated all of our vendor js files with `npm run vendor` (we should probably update our docs/script to include it, looks like it was lost after the docker update https://github.com/readthedocs/readthedocs.org/pull/9861#discussion_r1061894753)

stsewd commented on readthedocs/doc_builder/backends/mkdocs.py:
```diff
@@ -278,7 +278,9 @@ def generate_rtd_data(self, docs_dir, mkdocs_config):
             ),
-            'user_analytics_code': analytics_code,
+            "user_analytics_code": analytics_code,
+            "proxied_static_path": self.project.proxied_static_path,
+            "proxied_api_host": self.project.proxied_api_host,
```
We were missing including these for mkdocs, but it's only relevant when using a path different from `_/`, so

benjaoming reviewed (approved):
Solid work @santos22 :100:  I'm wondering if there's any adjustments that we'd want to make in the Sphinx 6 post https://github.com/readthedocs/blog/pull/201

On top of my head, I think it's great that this change ensures that nothing breaks - both our own Fly Out menu and whatever people have going on in their themes and projects can continue to use jQuery.

However, I'm a bit :-1: on documenting jQuery as something that's provided, hence also mentioning this in a blog post. People might use it, and then it gets even harder to remove.. when we want to deprecate it at some point.

I suppose that `proxied_static_path` is a heavily cached location?

agjohnson reviewed (approved):
This is a great fix for now :+1: 

I'll definitely want to remove this dependency as we break apart the theme and flyout JS. There definitely shouldn't be theme specific JS in the flyout JS and there shouldn't be flyout specific JS in the theme.
